File: server/app/avatar.py
"""Marque du joueur : une couleur, éventuellement un symbole.

Deux axes choisis par le joueur — l'anneau or/argent/bronze du top 3, lui, se gagne et
n'est pas stocké (il se déduit du classement à l'affichage). L'or, l'argent et le bronze
sont volontairement absents de la palette : un anneau doré ne doit vouloir dire qu'une
seule chose. Module pur, sans accès base, comme `elo.py`.

Les noms sont ceux du design system (tokens `--color-av-*` et jeu d'icônes maison) — toute
valeur qui n'est pas dans ces deux listes est refusée par `schemas.AvatarIn`.
"""

COLORS = ("citron", "jade", "lagon", "violet", "rose", "abricot")
SYMBOLS = ("trophee", "vie", "serie", "minuteur", "jouer", "clavier")


def default_color(username: str) -> str:
    """Couleur attribuée d'office à l'inscription, dérivée du pseudo.

    Déterministe : le même pseudo donne toujours la même couleur, ce qui rend un salon
    varié sans que personne n'ait eu à régler quoi que ce soit. Le joueur en change quand
    il veut. Hachage explicite (pas `hash()`, randomisé d'un process à l'autre).
    """
    h = 0
    for ch in username:
        h = (h * 31 + ord(ch)) & 0xFFFFFFFF
    return COLORS[h % len(COLORS)]

File: client/src/lib/avatar.ts
/** Marque du joueur — miroir client de `server/app/avatar.py` (les deux listes doivent
 * rester identiques : le serveur refuse toute valeur hors palette). La couleur est choisie
 * par le joueur, le symbole remplace ses initiales, et l'anneau or/argent/bronze du top 3
 * ne se choisit pas — il se déduit du classement à l'affichage. */

export const AVATAR_COLORS = ['citron', 'jade', 'lagon', 'violet', 'rose', 'abricot'] as const
export type AvatarColor = (typeof AVATAR_COLORS)[number]

export const AVATAR_SYMBOLS = ['trophee', 'vie', 'serie', 'minuteur', 'jouer', 'clavier'] as const
export type AvatarSymbol = (typeof AVATAR_SYMBOLS)[number]

const COLOR_SET: readonly string[] = AVATAR_COLORS

/** Token de couleur d'un avatar. Retombe sur citron pour une session d'avant l'avatar
 * (localStorage) ou une valeur inconnue — jamais de couleur en dur ailleurs. */
export function avatarColorVar(color?: string | null): string {
  return `var(--color-av-${color && COLOR_SET.includes(color) ? color : 'citron'})`
}

export const AVATAR_COLOR_LABELS: Record<AvatarColor, string> = {
  citron: 'Citron',
  jade: 'Jade',
  lagon: 'Lagon',
  violet: 'Violet',
  rose: 'Rose',
  abricot: 'Abricot',
}

export const AVATAR_SYMBOL_LABELS: Record<AvatarSymbol, string> = {
  trophee: 'Trophée',
  vie: 'Cœur',
  serie: 'Éclair',
  minuteur: 'Minuteur',
  jouer: 'Lecture',
  clavier: 'Clavier',
}
